File: elixir/rotational-cipher/rotational_cipher.exs
defmodule RotationalCipher do
  @doc """
  Given a plaintext and amount to shift by, return a rotated string.

  Example:
  iex> RotationalCipher.rotate("Attack at dawn", 13)
  "Nggnpx ng qnja"
  """
  @spec rotate(text :: String.t(), shift :: integer) :: String.t()
  def rotate(text, shift \\ 13) when is_binary(text) and is_integer(shift) do
    for <<char <- text>>, into: "", do: <<do_rotate(char, shift)>>
  end

  defp do_rotate(char, shift) when char in ?a..?z do
    Integer.mod(char - ?a + shift, 26) + ?a
  end

  defp do_rotate(char, shift) when char in ?A..?Z do
    Integer.mod(char - ?A + shift, 26) + ?A
  end

  defp do_rotate(char, _shift) do
    char
  end
end

File: ruby/rotational-cipher/rotational_cipher.rb
module RotationalCipher
  UPPERCASE_ALPHABET = (?A..?Z).to_a
  LOWERCASE_ALPHABET = UPPERCASE_ALPHABET.map(&:downcase)

  def self.rotate(plaintext, shift)
    plaintext.tr(
      (UPPERCASE_ALPHABET + LOWERCASE_ALPHABET).join,
      (UPPERCASE_ALPHABET.rotate(shift) + LOWERCASE_ALPHABET.rotate(shift)).join
    )
  end
end

module BookKeeping
  VERSION = 1
end
